pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java

```diff
@@ -169,6 +186,30 @@ private void start() {
 
     @Override
     CompletableFuture<MessageId> internalSendWithTxnAsync(Message<?> message, Transaction txn) {
+        int partition = routerPolicy.choosePartition(message, topicMetadata);
+        checkArgument(partition >= 0 && partition < topicMetadata.numPartitions(),
+                "Illegal partition index chosen by the message routing policy: " + partition);
+
+        if (!producers.containsKey(partition)) {
+            final State createState = createProducer(partition).handle((prod, createException) -> {
+                if (createException != null) {
+                    log.error("[{}] Could not create internal producer. partitionIndex: {}", topic, partition,
+                            createException);
+                    try {
+                        producers.remove(partition).close();
```

Comment:
We are not sure that we are removing the newly create producer here.

Reply:
The  reason for it is to recreate internal producer when producer creation is failed. This behavior is tested [here](https://github.com/apache/pulsar/pull/10279/files#diff-eee943fd1468c4cbe8922b01a487f153480b9ccbaa9606f114bb0bb373ba5859R493-R546).

Reply:
I mean that we should test the result for "producers.remove(partition)" for equality ('==') with `prod`